gratipay/models/exchange_route.py

```diff
@@ -69,20 +67,35 @@ def insert(cls, participant, network, address, error='', fee_cap=None, cursor=No
     def invalidate(self):
         if self.network == 'braintree-cc':
             braintree.PaymentMethod.delete(self.address)
-
-        # For Paypal, we remove the record entirely to prevent
-        # an integrity error if the user tries to add the route again
-        if self.network == 'paypal':
-            # XXX This doesn't sound right. Doesn't this corrupt history pages?
-            self.db.run("DELETE FROM exchange_routes WHERE id=%s", (self.id,))
-        else:
-            self.update_error('invalidated')
+        with self.db.get_cursor() as cursor:
+            self.db.run("UPDATE exchange_routes SET is_deleted=true WHERE id=%s", (self.id,))
+            payload = dict( id=self.participant.id
+                          , exchange_route=self.id
+                          , action='invalidate route'
```

Comment:
not quite familiar with this 'action' field. I imagine there are a set of actions, each of which should be constants. IMO they are better to be defined in one place, but feel free to ignore if it involves too much refactoring.